plugin/src/main/java/org/opensearch/ml/action/models/DeleteModelTransportAction.java

```diff
@@ -214,47 +214,101 @@ private void returnFailure(BulkByScrollResponse response, String modelId, Action
         actionListener.onFailure(new OpenSearchStatusException(errorMessage, RestStatus.INTERNAL_SERVER_ERROR));
     }
 
-    private void deleteModel(String modelId, ActionListener<DeleteResponse> actionListener) {
-        DeleteRequest deleteRequest = new DeleteRequest(ML_MODEL_INDEX, modelId);
-        client.delete(deleteRequest, new ActionListener<DeleteResponse>() {
+    private void deleteModel(String modelId, FunctionName functionName, ActionListener<DeleteResponse> actionListener) {
+        DeleteRequest deleteRequest = new DeleteRequest(ML_MODEL_INDEX, modelId).setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
+        client.delete(deleteRequest, new ActionListener<>() {
             @Override
             public void onResponse(DeleteResponse deleteResponse) {
-                deleteModelChunks(modelId, deleteResponse, actionListener);
-                deleteController(modelId);
+                deleteModelChunksAndController(functionName, actionListener, modelId, deleteResponse);
             }
 
             @Override
             public void onFailure(Exception e) {
-                log.error("Failed to delete model meta data for model: " + modelId, e);
                 if (e instanceof ResourceNotFoundException) {
-                    deleteModelChunks(modelId, null, actionListener);
-                    deleteController(modelId);
+                    deleteModelChunksAndController(functionName, actionListener, modelId, null);
+                } else {
+                    log.error("Failed to delete model meta data for model, please try again: " + modelId, e);
+                    actionListener.onFailure(e);
                 }
-                actionListener.onFailure(e);
             }
         });
     }
 
+    private void deleteModelChunksAndController(
+        FunctionName functionName,
+        ActionListener<DeleteResponse> actionListener,
+        String modelId,
+        DeleteResponse deleteResponse
+    ) {
+        if (FunctionName.REMOTE != functionName) {
+            CountDownLatch countDownLatch = new CountDownLatch(2);
+            AtomicBoolean bothDeleted = new AtomicBoolean(true);
+            ActionListener<Boolean> countDownActionListener = ActionListener.wrap(b -> {
+                countDownLatch.countDown();
+                bothDeleted.compareAndSet(true, b);
+                if (countDownLatch.getCount() == 0) {
+                    if (bothDeleted.get()) {
+                        log.debug("model chunks and model controller for model {} deleted successfully", modelId);
+                        if (deleteResponse != null) {
+                            actionListener.onResponse(deleteResponse);
```

Comment:
Seems `deleteResponse` could be null, will it cause any error? What information user can see for null?

Reply:
Only when model not found, deleteResponse will be null, in this case we try to delete model chunks and model controller, and then return 400 "model not found" to user.